### PR TITLE
Improve handling of external grader and workspace arguments

### PR DESCRIPTION
**Pull request by jonatanschroeder**
The current external grader entrypoint processing currently has some inconsistent handling of arguments: the entrypoint is split, but some parts of the code assume it does no have arguments. This PR changes the functionality to explicitly support arguments.

In addition to this, both for external grader entrypoints and workspace arguments, it also:
* improves support for proper splitting of arguments with quotes/escaping, by using proper shlex splitting;
* adds support for setting these attributes in the JSON file as arrays.

It also fixes the code that adds executable permissions to the entrypoint, by ensuring it is only given to files that are mounted in the /grade volume (docs are also updated to make this clear). Additionally, it provides consistency in the dev vs grader-host implementation of the external grader to handle cases where the entrypoint is in a /grade directory other than serverFilesCourse (e.g., /grade/tests). 

* [x] CI tests;
* [x] Example course question.

### CODE REVIEW

File: apps/grader-host/src/index.ts [hunk 1]
```diff
@@ -324,8 +325,15 @@ async function initFiles(context: Context) {
     logger.info('Unzipping files');
     await execa('tar', ['-xf', jobArchiveFile.path, '-C', jobDirectory.path]);
 
+    // TODO: AFAICT this is to handle entrypoints that start with /grade/ or
+    // equivalent (e.g., in /grade/serverFilesCourse or /grade/tests). If that
+    // is the case, this code should only run if the entrypoint is indeed in
+    // that directory.
```
Comment (jonatanschroeder) [inline apps/grader-host/src/index.ts:331]: @nwalters512 I believe the correct handling here is to check if the entrypoint (or more precisely the first token of the entrypoint) starts with `/grade`. Can you double-check if this makes sense?

File: apps/prairielearn/src/lib/externalGraderLocal.ts [hunk 2]
```diff
@@ -73,12 +74,19 @@ export class ExternalGraderLocal {
...
-        // we actually run it.
+        // If it is living in the docker container then we don't have access to
+        // it before we actually run it.
+        //
+        // TODO: This code differs from grader-host, in that it only runs for
+        // serverFilesCourse, while grader-host runs for all entrypoints. There
+        // should be consistency in this case to address cases where the
+        // entrypoint is, e.g., in /grade/tests instead. It also assumes that
+        // `serverFilesCourse` is prefixed by `/grade/`, which is a reasonable
+        // assumption, but should probably be checked.
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/externalGraderLocal.ts:85]: @nwalters512 similar question, but in this case this should replace the `serverFilesCourse` check with a `/grade` test instead. Can you again double check?

File: apps/grader-host/src/index.ts (hunk 1 above)
Comment (nwalters512) [reply apps/grader-host/src/index.ts:331]: That sounds right, yeah.

File: apps/prairielearn/src/lib/externalGraderLocal.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/lib/externalGraderLocal.ts:85]: This sounds about right, yeah. I think this was specifically trying to handle `/grade/serverFilesCourse`, but it seems sensible to generalize this to `/grade/...`.

File: apps/grader-host/src/index.ts (hunk 1 above)
Comment (nwalters512) [reply apps/grader-host/src/index.ts:331]: Note that we probably want to ensure that we do something sensible for an entrypoint like `/grade/../../../other-thing` (that is, guard against directory traversals).

File: apps/prairielearn/src/lib/externalGraderLocal.ts (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/externalGraderLocal.ts:85]: Which seems like it was a problem in the past...
https://github.com/PrairieLearn/PrairieLearn/blob/c74134815f90bbc8144489d44c28b44a95f46541/docs/c-grader/index.md#L26-L31

File: apps/grader-host/src/index.ts (hunk 1 above)
Comment (jonatanschroeder) [reply apps/grader-host/src/index.ts]: Done in 4b6bccca9dc87cba8c664ff70211fff739e0e3b0.

File: apps/prairielearn/src/lib/externalGraderLocal.ts (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/externalGraderLocal.ts]: Done in 4b6bccca9dc87cba8c664ff70211fff739e0e3b0.

File: apps/prairielearn/src/lib/server-jobs.ts [hunk 3]
```diff
@@ -117,7 +118,7 @@ class ServerJobImpl implements ServerJob, ServerJobExecutor {
     args: string[] = [],
     options: ServerJobExecOptions,
   ): Promise<ServerJobResult> {
-    this.addToOutput(chalk.blueBright(`Command: ${file} ${args.join(' ')}\n`));
+    this.addToOutput(chalk.blueBright(`Command: ${shlex.join([file, ...args])}\n`));
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/server-jobs.ts:121]: This is not directly related to the other code, but since shlex has been added might as well use it here... Note that the previous use would not be an error, since this is just logging, but now the logging will properly handle cases with e.g., spaces in arguments.

File: apps/grader-host/src/index.ts [hunk 4]
```diff
@@ -397,7 +406,7 @@ async function runJob(context: Context, receivedTime: Date, tempDir: string) {
           },
         ],
       },
-      Entrypoint: entrypoint.split(' '),
+      Entrypoint: shlex.split(entrypoint),
```
Comment (jonatanschroeder) [inline apps/grader-host/src/index.ts:409]: If someone is able to test the grader host updated functionality, that would be great. This can be done with the new alpine question in testCourse (`externalGrade/entrypointArgs`), which has an entrypoint with quoted arguments and a grade.py file that is purposefully not executable in the repo.

**Review by nwalters512 — APPROVED** [review]
This looks more or less good to go. I deployed `grader-host` to our staging environment and tried a copy of the `externalGrade/entrypointArgs` question with both of the following:

```
"entrypoint": ["/grade/tests/grade.sh", "answer with space"],
"entrypoint": "/grade/tests/grade.sh 'answer with space'",
```

In each case, things worked as expected!

Approving, but I also left a few minor comments.

File: exampleCourse/questions/demo/workspace/xtermjsPython/info.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "uuid": "b62a4383-cbc1-4a10-b058-2fab814e6c17",
+  "title": "Workspace demo: Xterm.js running Python",
+  "topic": "Workspace",
+  "tags": ["su20", "nnytko2", "tyang15"],
+  "type": "v3",
+  "singleVariant": true,
+  "workspaceOptions": {
+    "comment": { "why": "testing comments in workspace blocks", "also": "testing object comments" },
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply exampleCourse/questions/demo/workspace/xtermjsPython/info.json]: Done in 4427e4868fe4908a5075ffeb1ba6d8b097e74b85.

File: exampleCourse/questions/demo/workspace/xtermjsPython/info.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "uuid": "b62a4383-cbc1-4a10-b058-2fab814e6c17",
+  "title": "Workspace demo: Xterm.js running Python",
+  "topic": "Workspace",
+  "tags": ["su20", "nnytko2", "tyang15"],
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply exampleCourse/questions/demo/workspace/xtermjsPython/info.json]: Done in 4427e4868fe4908a5075ffeb1ba6d8b097e74b85.

File: apps/prairielearn/src/lib/externalGraderLocal.ts [hunk 7]
```diff
@@ -71,17 +73,23 @@ export class ExternalGraderLocal {
...
-        // If it is living in the docker container then we don't have access to it before
-        // we actually run it.
-        try {
-          await execa('chmod', [
-            '+x',
-            path.join(dir, question.external_grading_entrypoint.slice(6)),
-          ]);
-        } catch {
-          logger.error('Could not make file executable; continuing execution anyways');
+      if (question.external_grading_entrypoint) {
```
[the remark this answers is not in the document]
Comment (jonatanschroeder) [reply apps/prairielearn/src/lib/externalGraderLocal.ts]: Added #10940 for your comment. Also added a check and exception to avoid typecheck issues.

**Comment by jonatanschroeder** [comment]
I know it's approved, but I'd love a double-check on the changes for null entrypoint. Note that there is no expected case where entrypoint will be null in that scenario without also having a null image (which would have caused an error just before that case), but I decided to make it explicit to make typecheck happy and to make the behaviour clear in the code.

File: apps/prairielearn/src/lib/externalGraderLocal.ts [hunk 8]
```diff
@@ -64,21 +66,37 @@ export class ExternalGraderLocal {
...
 
+      if (question.external_grading_entrypoint == null) {
+        // The grader-host does not currently support a null entrypoint, and
+        // schema does not allow the entrypoint to be skipped if external
+        // grading options are provided. To avoid local code having a different
+        // behavior than production, we ensure that the local code throws an
+        // error if the entrypoint is null, which would also happen in
+        // production. A future implementation could allow the entrypoint to be
+        // skipped to use the default entrypoint from the image, but this would
+        // require equivalent changes to the grader-host.
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/externalGraderLocal.ts:77]: Might as well link to the issue since we have it.

```suggestion
        // The grader-host does not currently support a null entrypoint, and
        // schema does not allow the entrypoint to be skipped if external
        // grading options are provided. To avoid local code having a different
        // behavior than production, we ensure that the local code throws an
        // error if the entrypoint is null, which would also happen in
        // production. A future implementation could allow the entrypoint to be
        // skipped to use the default entrypoint from the image, but this would
        // require equivalent changes to the grader-host; see
        // https://github.com/PrairieLearn/PrairieLearn/issues/10940.
```

**Review by nwalters512 — APPROVED** [review]
LGTM!
